petab/visualize/helper_functions.py

```diff
@@ -390,56 +396,208 @@ def get_default_vis_specs(
+                    column_entries.append(
+                        vis_spec[select_conditions].loc[:, column].values[0])
+            else:
+                for plot_id in columns_dict[PLOT_ID]:
+                    select_conditions = vis_spec[PLOT_ID] == plot_id
+                    column_entries.append(
+                        vis_spec[select_conditions].loc[:, column].values[0])
+            columns_dict[column] = column_entries
+    vis_spec = pd.DataFrame(columns_dict)
+    return vis_spec
+
+
+def create_or_update_vis_spec(
+        exp_data: pd.DataFrame,
+        exp_conditions: pd.DataFrame,
+        vis_spec: Optional[pd.DataFrame] = None,
+        dataset_id_list: Optional[List[IdsList]] = None,
+        sim_cond_id_list: Optional[List[IdsList]] = None,
+        sim_cond_num_list: Optional[List[NumList]] = None,
+        observable_id_list: Optional[List[IdsList]] = None,
+        observable_num_list: Optional[List[NumList]] = None,
+        plotted_noise: Optional[str] = MEAN_AND_SD):
+    """
+    Helper function for plotting data and simulations, which updates vis_spec
+    file if necessary or creates a default visualization table and
+    updates/creates DATASET_ID column of exp_data. As a result, a visualization
+    specification file exists with columns PLOT_ID, DATASET_ID, Y_VALUES and
+    LEGEND_ENTRY
+
+    Returns:
+        A tuple of visualization specification DataFrame and experimental
+        DataFrame.
+    """
+    if vis_spec is None:
+        # create dataframe
+        exp_data, columns_dict = \
+            get_vis_spec_dependent_columns_dict(exp_data,
+                                                exp_conditions,
+                                                dataset_id_list,
+                                                sim_cond_id_list,
+                                                sim_cond_num_list,
+                                                observable_id_list,
+                                                observable_num_list)
+        vis_spec = pd.DataFrame(columns_dict)
+    else:
+        # TODO: do validation
+        # so, plotid is definitely there
+        if DATASET_ID not in vis_spec.columns:
+            if Y_VALUES in vis_spec.columns:
+                plot_id_list = np.unique(vis_spec[PLOT_ID])
+
+                observable_id_list = [vis_spec[vis_spec[PLOT_ID] ==
+                                               plot_id].loc[:, Y_VALUES].values
+                                      for plot_id in plot_id_list]
+                exp_data, columns_dict = \
+                    get_vis_spec_dependent_columns_dict(
+                        exp_data,
+                        exp_conditions,
+                        observable_id_list=observable_id_list)
+            else:
+                # PLOT_ID is there, but NOT DATASET_ID and not Y_VALUES,
+                # but potentially some settings.
+                # TODO: multiple plotids with diff settings
```

Comment:
this does not work with multiple plotIDs right now?
If it does not raise warning. If it does remove comment

Reply:
vis_spec like this won't work right now

plotId | xOffset | yScale
-- | -- | --
plot1 | 100 | log
plot2 | 50 | log
plot3 | 0 | lin

I don't know if anyone would want to use such vis_spec. Potentially could be tried to plot everything with different settings.